i2c/mcp9808_i2c/mcp9808_i2c.c

```diff
@@ -0,0 +1,153 @@
+
+    // Check flags and raise alerts accordingly 
+    if ((upper_byte & 0x40) == 0x40){ //TA > TUPPER
+        printf("Temperature is above the upper temperature limit.\n");
+    }
+    if ((upper_byte & 0x20) == 0x20) { //TA < TLOWER
+        printf("Temperature is below the lower temperature limit.\n");
+    }
+    if ((upper_byte & 0x80) == 0x80) { //TA > TCRIT
+        printf("Temperature is above the critical temperature limit.\n");
+    }
+}
+
+float mcp9808_convert_temp(uint8_t upper_byte, uint8_t lower_byte) {
+
+    float temperature;
+
+
+    //Check if TA <= 0°C and convert to denary accordingly
+    if ((upper_byte & 0x10) == 0x10) { 
+        upper_byte = upper_byte & 0x0F; 
+        temperature = 256 - (((float)upper_byte * 16) + ((float)lower_byte / 16));
+    } else {
+        temperature = (((float)upper_byte * 16) + ((float)lower_byte / 16));
+  
+    }
+    return temperature;
+}
+
+
+void mcp9808_set_limits() {
+    
+    //Set an upper limit of 30°C for the temperature
+    uint8_t upper_temp_msb = 0x01;
+    uint8_t upper_temp_lsb = 0xE0;
+
+    //Set a lower limit of 20°C for the temperature
+    uint8_t lower_temp_msb = 0x01;
+    uint8_t lower_temp_lsb = 0x40;
+
+    //Set a critical limit of 40°C for the temperature
+    uint8_t crit_temp_msb = 0x02;
+    uint8_t crit_temp_lsb = 0x80;
+
+    uint8_t buf[3];
+    buf[0] = REG_TEMP_UPPER;
+    buf[1] = upper_temp_msb;
+    buf[2] = upper_temp_lsb; 
+    i2c_write_blocking(i2c_default, ADDRESS, buf, 3, false);
+
+    buf[0] = REG_TEMP_LOWER;
+    buf[1] = lower_temp_msb;
+    buf[2] = lower_temp_lsb; 
+    i2c_write_blocking(i2c_default, ADDRESS, buf, 3, false);
+
+    buf[0] = REG_TEMP_CRIT;
+    buf[1] = crit_temp_msb;
+    buf[2] = crit_temp_lsb; ;
+    i2c_write_blocking(i2c_default, ADDRESS, buf, 3, false);
+}
+
+
+int main() {
+
+    stdio_init_all();
+
+#if !defined(i2c_default) || !defined(PICO_DEFAULT_I2C_SDA_PIN) || !defined(PICO_DEFAULT_I2C_SCL_PIN)
+    #warning i2c/mcp9808_i2c example requires a board with I2C pins
+    puts("Default I2C pins were not defined");
+#else
+    printf("Hello, MCP9808! Reading raw data from registers...\n");
+
+    // This example will use I2C0 on the default SDA and SCL pins (4, 5 on a Pico)
+    i2c_init(i2c_default, 400 * 1000);
+    gpio_set_function(PICO_DEFAULT_I2C_SDA_PIN, GPIO_FUNC_I2C);
+    gpio_set_function(PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C);
+    gpio_pull_up(PICO_DEFAULT_I2C_SDA_PIN);
+    gpio_pull_up(PICO_DEFAULT_I2C_SCL_PIN);
+    // Make the I2C pins available to picotool
+    bi_decl(bi_2pins_with_func(PICO_DEFAULT_I2C_SDA_PIN, PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C));
+#endif
+
+    mcp9808_set_limits();
+    
+    uint8_t buf[2];
+    uint16_t upper_byte;
+    uint16_t lower_byte;
+
+    float temperature;
+
+    while (1) {
+        // Start reading ambient temperature register for 2 bytes
+        i2c_write_blocking(i2c_default, ADDRESS & WRITE_MODE, &REG_TEMP_AMB, 1, true);

```

Comment:
The SDK adds these bits to the end for us. `i2c_write_blocking` expects a 7-bit address, the code then shifts that to the left by 1 bit and adds the appropriate read/write bit.

Reply:
ie. we can get rid of this of write mode, read mode and the logical ands. on line 104 for example, this isn't done and the code still works, so proof that it's not needed :P

Reply:
Oh nice! Quite handy, gotten rid of all that then

Reply:
> The SDK adds these bits to the end for us. i2c_write_blocking expects a 7-bit address, the code then shifts that to the left by 1 bit and adds the appropriate read/write bit.

 @matiasilva Maybe you want to send a PR to the SDK clarifying the documentation (in the doxygen) there, to prevent other users running into the same trap? :thinking:

Reply:
Good call, see https://github.com/raspberrypi/pico-sdk/pull/520 :D